cpp/src/parquet/page_index.cc

```diff
@@ -0,0 +1,254 @@
+            column_index.__isset.null_counts, column_index.null_counts) {}
+
+  int64_t num_pages() const override { return static_cast<int64_t>(null_pages_.size()); }
+
+  bool is_null_page(int64_t page_id) const override {
+    if (page_id >= num_pages()) {
+      throw ParquetException("Page index is out of bound");
+    }
+    return null_pages_[page_id];
+  }
+
+  BoundaryOrder boundary_order() const override { return boundary_order_; }
+
+  bool has_null_counts() const override { return has_null_count_; }
+
+  int64_t null_count(int64_t page_id) const override {
+    if (page_id >= num_pages()) {
+      throw ParquetException("Page index is out of bound");
+    }
+    return null_counts_[page_id];
+  }
+
+  T min_value(int64_t page_id) const override {
+    return min_values_[GetMinMaxSlot(page_id)];
+  }
+
+  T max_value(int64_t page_id) const override {
+    return max_values_[GetMinMaxSlot(page_id)];
+  }
+
+  const std::string& encoded_min(int64_t page_id) const override {
+    if (page_id >= num_pages()) {
+      throw ParquetException("Page index is out of bound");
+    }
+    return encoded_min_values_[page_id];
+  }
+
+  const std::string& encoded_max(int64_t page_id) const override {
+    if (page_id >= num_pages()) {
+      throw ParquetException("Page index is out of bound");
+    }
+    return encoded_max_values_[page_id];
+  }
+
+  const std::vector<bool>& null_pages() const override { return null_pages_; }
+
+  const std::vector<int64_t>& null_counts() const override { return null_counts_; }
+
+  const std::vector<T>& min_values() const override { return min_values_; }
+
+  const std::vector<T>& max_values() const override { return max_values_; }
+
+  std::vector<int64_t> GetValidPageIndices() const override {
+    std::vector<int64_t> valid_page_indices;
+    std::for_each(page_indexes_.cbegin(), page_indexes_.cend(),
+                  [&](const std::pair<size_t, size_t>& v) {
+                    valid_page_indices.push_back(v.first);
+                  });
+    return valid_page_indices;
+  }
+
+ private:
+  size_t GetMinMaxSlot(int64_t page_id) const {
+    if (page_id >= static_cast<int64_t>(null_pages_.size())) {
+      throw ParquetException("Page index is out of bound");
+    }
+    if (null_pages_[page_id]) {
+      throw ParquetException("Cannot get min/max value of null page");
+    }
+    auto iter = page_indexes_.find(page_id);
+    if (iter == page_indexes_.cend()) {
+      throw ParquetException("min/max value is unavailable");
+    }
+    return iter->second;
+  }
+
+  /// Values that are copied directly from the thrift message.
+  std::vector<bool> null_pages_;
+  std::vector<std::string> encoded_min_values_;
+  std::vector<std::string> encoded_max_values_;
+  BoundaryOrder boundary_order_;
+  bool has_null_count_;
+  std::vector<int64_t> null_counts_;
+
+  /// page_id -> slot_id in the buffer of min_values_ & max_values_
+  std::map<size_t, size_t> page_indexes_;
```

Comment:
I guess there is a trade-off but is there  a reason this data-structure is used instead of a std::vector<size_t>, that stores -1 for slots that are all null and the corresponding slot value otherwise?

Also any reason for map vs unordered_map?

Reply:
The trade-off is that we have the chance to save some memory if there is any null page. Using `map` instead of `unordered_map` because GetValidPageIndices() is able to generate ordered page indices simply by iterating the map.